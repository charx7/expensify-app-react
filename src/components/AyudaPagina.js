// Importaciones de React
import React from 'react';
import ReactDOM from 'react-dom';

// Componente de Editar Gasto
const AyudaPagina = () => (
    <div>Aiudaaa</div>
);

export default AyudaPagina;