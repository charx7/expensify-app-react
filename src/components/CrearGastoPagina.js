// Importaciones de React
import React from 'react';
import ReactDOM from 'react-dom';

// Componente de Crear Gasto
const CrearGastoPagina = () => (
    <div>Crear Gasto</div>
);

export default CrearGastoPagina;