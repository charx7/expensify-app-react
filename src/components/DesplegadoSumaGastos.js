import React from 'react';
import { connect } from 'react-redux';
import gastosSeleccionados from '../selectores/selectorGastos'; // Importaciones del filtro
import sumaGastosVisibles from '../selectores/gastosTotales'; // Importacion del modulo que hace la suma de los gastos seleccionados
import numeral from 'numeral'; // Importaciones de node_modules para formato de montos

// Creamos un componente de clase 
const DesplegadoSumaGastos = (props) => (
    <div>
        {/* Aca va el reesultado de la suma de los gasot*/}
        <p>
            Viendo {props.gastosVisibles.length} gasto{props.gastosVisibles.length == 1 ? ' ' : 's '} 
            sumando: {numeral(sumaGastosVisibles(props.gastosVisibles)/100).format('$0,0.[00]')}
        </p>
    </div>
);

// Funcion de mapeo del estado a los props que se pasan al componente
const mapeoEstadoAProps = (estado) => {
    return {
        // Pasa un prop al componente ListadeGastos basado en el mapeo del almacen y filtrados con la funcion de selectorGastos
        gastosVisibles: gastosSeleccionados(estado.gastos, estado.filtros)
    }
}

// Exportaciones con conexion al almacen
export default connect(mapeoEstadoAProps)(DesplegadoSumaGastos);