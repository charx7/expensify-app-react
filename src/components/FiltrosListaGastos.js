import "react-dates/initialize"; // Importacion nueva necesaria para que sirva el calendario
import 'react-dates/lib/css/_datepicker.css' // Importacion del CSS
import React from 'react';
import { connect } from 'react-redux';
import { setFiltroTexto, filtroOrdenarPorFecha, filtroOrdenarPorMonto,
         filtroFechaInicio, filtroFechaFinal } from '../acciones/filtros'; // importamos la accion de modificar el texto del filtro
import { DateRangePicker } from 'react-dates'; // Importaciones para el picker de las fechas

// Componente de clase porque usaremos un estado para manejar el calendario
class FiltrosListaGasto extends React.Component{
    state = {
        focusCalendario: null
    };

    // Metodo encargado de la logica del cambio de fechas
    onDateChange = ( { startDate, endDate } ) => {
        // Llamamos a las acciones que modifican las fechas de los filtros en el almacen
        this.props.dispatch(filtroFechaInicio(startDate));
        this.props.dispatch(filtroFechaFinal(endDate));
    };

    // Metodo de cambio de foco del calendario picker
    onFocusChange = (focusCalendario) => {
        this.setState(() => ({focusCalendario: focusCalendario}));
    };

    render() {
        return (
            <div>
                <input 
                    type='text' 
                    value={this.props.filtros.texto}
                    /* Metodo que ejecuta una funcion cuando se hace submit al input en este caso cambia los filtros*/
                    onChange = { (e) => {
                        this.props.dispatch(setFiltroTexto(e.target.value));
                        console.log(e.target.value)
                    }}
                />
                <select value={this.props.filtros.opcionOrdenar} onChange={ (e) => {
                    if ( e.target.value == 'fechaSelector') {
                        this.props.dispatch(filtroOrdenarPorFecha());
                    } else {
                        this.props.dispatch(filtroOrdenarPorMonto());
                    }
                }}>
                    <option value='fechaSelector'>Fecha</option>
                    <option value='montoSelector'>Monto</option>
                </select>
                <DateRangePicker
                    startDate     = {this.props.filtros.fechaInicio}
                    endDate       = {this.props.filtros.fechaFinal}
                    onDatesChange = {this.onDateChange}
                    focusedInput  = {this.state.focusCalendario}
                    onFocusChange = {this.onFocusChange}
                    showClearDates = {true}
                    numberOfMonths = {1}
                    isOutsideRange = {() => false}
                    startDateId   = {'startDate'}
                    endDateId     = {'endDate'}
                />
            </div>
        );
    }
}

// Funcion de mapeo de los elementos del almacen a los props 
const mapeoEstadoAProps = (estado) => {
    return {
        filtros: estado.filtros
    };
};

// Exportaciones con conexion al almacen
export default connect(mapeoEstadoAProps)(FiltrosListaGasto);