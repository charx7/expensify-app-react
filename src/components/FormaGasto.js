// Archivo que contiene la logica de la forma para editar o crear un gasto
import "react-dates/initialize"; // Importacion nueva necesaria para que sirva el calendario
import React from  'react';
import moment from 'moment'; // Importacion de momentos
import { SingleDatePicker } from 'react-dates'; // Importacion de React Dates
import 'react-dates/lib/css/_datepicker.css' // Importacion del CSS

// Creamos un objeto de la libreria moment
const now = moment();
console.log(now.format('MMM Do, YYYY'));

// Componente de clase
export default class FormaGasto extends React.Component{
    // Override al constructor
    constructor(props) {
        // Llamamos al constructor de la clase padre
        super(props);
        // Estado del componente que mantiene track de los cambios de los inputs y el textarea
        this.state = {
            descripcion: props.gastoEditar ? props.gastoEditar.descripcion : '',
            nota: props.gastoEditar ? props.gastoEditar.nota : '',
            monto: props.gastoEditar ? (props.gastoEditar.monto /100).toString() : '',
            creadoEn: props.gastoEditar ? moment(props.gastoEditar.creadoEn) : moment(),
            calendarFocused: false,
            error: ''
        }; 
    }
    // Metodo que se encarga de cambiar el estado de descripcion del componente
    cambioDescripcion = (e) => {
        // Recupera el valor del input de texto de descripcion
        const descripcionForma = e.target.value;
        // Cambia el estado del componente con el objeto del input de descripcion
        this.setState( () => ({
            descripcion: descripcionForma
        }));
    }
    // Metodo que se encarga del cambio el estado del componente de nota
    cambioNota = (e) => {
        // Jala el valor puesta en el area de texto de la forma
        const notaForma = e.target.value;
        // Cambio del estado del componente segun el valor del area de texto
        this.setState( () => ({ nota: notaForma }));
    }
    // Metodo que se encarga del cambio de estado de monto
    cambioMonto = (e) => {
        // Jalamos el valor del input
        const montoForma = e.target.value;
        // Logica de verificacion del input el non-sense es una regular-expression que se encarga de verificar si es un
        // numero con 2 decimales maximo
        if( !montoForma || montoForma.match( /^\d{1,}(\.\d{0,2})?$/ )) {
            // Cambio del estado
            this.setState( () => ({ monto: montoForma}));
        }
    };
    // Metodo que se encarga de manipular el estado de Fecha segun el calendario chevere de la libreria 3rd party
    cambioFecha = (creadoEnForma) => {
        if(creadoEnForma) {
            this.setState(() => ({ creadoEn: creadoEnForma }));
        }
    };
    // Metodo de cambio de focus cuando se manipula el calendarito chevere
    enCambioCalendarFocused = ({ focused }) => {
        this.setState( () => ({ calendarFocused: focused }));
    };
    // metodo para submit de la forma
    submitForma = (e) => {
        // Para que no ocurra un full refresh cuando el user le da click
        e.preventDefault();
        if(this.state.monto && this.state.descripcion) {
            // Limpiar el error
            console.log('Submited!');
            this.setState( () => ({ error: '' }));
            // Llamamos al metodo que pasamos como prop que se encarga de agregar los datos al almacen
            // Y le pasamos el objeto con los valores del estado
            this.props.onSubmit({
                descripcion: this.state.descripcion,
                monto: parseFloat(this.state.monto,10) * 100,
                creadoEn: this.state.creadoEn.valueOf(),
                nota: this.state.nota
            });
        } else {
            // Poners un error
            console.log('error');
            this.setState( () => ({ error: 'Error XD Porfis capture descripcion y monto.' }));
        }
    };
    // Rendereo del componente
    render () {
        return (
        <div>
            {this.state.error && <p>{this.state.error}</p>}
            <form onSubmit={this.submitForma}>
                <input type="text"
                    placeholder='descripcion'
                    // Hace que se autoseleccione la forma al renderear el componente
                    autoFocus
                    value = {this.state.descripcion}
                    onChange = {this.cambioDescripcion}
                 />
                 <input type="text"
                    placeholder='monto'
                    value = {this.state.monto}
                    onChange = {this.cambioMonto}
                  />
                <textarea 
                    placeholder='Agregar Nota para tu gasto (opcional)'
                    value = {this.state.nota}
                    onChange = {this.cambioNota}
                >
                </textarea>
                <button>Agregar Gasto</button>
            </form>
            {/* Importaciones del componente de Single date picker de react-dates*/}
            <SingleDatePicker
                // Props necesarios para que funcione el calendario
                date = {this.state.creadoEn}
                onDateChange = {this.cambioFecha}
                focused = {this.state.calendarFocused}
                onFocusChange = {this.enCambioCalendarFocused}
                numberOfMonths = {1}
                isOutsideRange = {(day) => false }
            />
        </div>
        )
    }
}

