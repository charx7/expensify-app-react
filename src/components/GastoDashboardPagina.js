// Importaciones de React
import React from 'react';
import ReactDOM from 'react-dom';
// Importando nuestro componente
import ListaGastos from './ListaGastos';
import FiltrosListaGasto from './FiltrosListaGastos'; // Importacion del componente de filtros

// Prueba
const GastoDashboardPagina = () => (
    <div>
        {/* Importamos el componente que se encarga de hacer set a los parametros del filtro*/}
        <FiltrosListaGasto/>
        {/* Rendereamos la lista de gastos del componente importado*/}
        <ListaGastos/>
    </div>
);

export default GastoDashboardPagina;