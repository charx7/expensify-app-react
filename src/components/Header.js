import React from 'react';
import {NavLink} from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Expensify App</h1>
        
        <NavLink to = "/" activeClassName = "esta-activo" exact={true}>
            DASHBOARD
        </NavLink>

        <NavLink to = "/crear" activeClassName = "esta-activo">
            Crear
        </NavLink>

        <NavLink to = "/ayuda" activeClassName = "esta-activo">
            Ayuda
        </NavLink>
    </header>
);

export default Header;