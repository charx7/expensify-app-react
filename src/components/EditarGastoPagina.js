// Importaciones de React
import React from 'react';
import ReactDOM from 'react-dom';
// Importacion de connect de React redux
import { connect } from 'react-redux';
import FormaGasto from  './FormaGasto'; // Importamos la forma de gasto
import { editarGasto, empiezaRemoverGasto, empiezaEditarGasto } from '../acciones/gastos'; // Importamos la accion de editar gasto

// Componente de Editar Gasto
const EditarGastoPagina = (props) => {
    return (
        <div>
            Editando el Gasto con ID de {props.match.params.id}
            <FormaGasto
                gastoEditar = {props.gasto} 
                onSubmit = {(gasto) => {
                    console.log('Editando ', gasto);
                    // Mandar la nueva info de gasto a la accion de editar
                    props.dispatch(empiezaEditarGasto(props.match.params.id, gasto));
                    // Redirigir a la pagina de dashboard
                    props.history.push('/');
                }}
            />
            <button 
                onClick = { () => {
                    // Llamado a la accion que remueve un gasto del almacen
                    props.dispatch(empiezaRemoverGasto({id: props.match.params.id}));
                    // Redirigir a la pagina de Dashboard
                    props.history.push('/');
                }}
            >
                Eliminar
            </button>
        </div>
    );
};

// Mapeo del estado del almacen a props
const mapeoEstadoAProps = (estado, props) => {
    return {
        // Buscamos en el arreglo de gastos del almacen para que nos devuelva el objeto con las propiedades indicadas
        gasto: estado.gastos.find( (gasto) => {
            // Verifica si el argumento pasado tiene el mismo id que el del almacen
            return gasto.id == props.match.params.id;
        })
    };
};

// Conexion con el Store
export default connect(mapeoEstadoAProps) (EditarGastoPagina);