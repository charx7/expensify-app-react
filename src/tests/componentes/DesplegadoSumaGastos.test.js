import React from 'react'; // Importacion de React
import { shallow } from 'enzyme'; // Libreria de enzyme para testear componentes de react
import { DesplegadoSumaGastos } from '../../components/DesplegadoSumaGastos'; // Importamos el componente de React a testear
import moment from 'moment';

const gastosPrueba1 = [{
    id: '1',
    descripcion: 'Gum',
    nota: '',
    monto: 195,
    creadoEn: 0
}]

const gastosPrueba2 = [{
    id: '1',
    descripcion: 'Gum',
    nota: '',
    monto: 195,
    creadoEn: 0 
}, {
    id: '2',
    descripcion: 'Renta',
    nota: '',
    monto: 1000,
    creadoEn: moment(0).subtract(4,'days').valueOf() 
}]

test('Deberia renderear bien el componente con un gasto vacio', () =>{
    // Lo que se va renderear al usuario
    const wrapper = shallow(<DesplegadoSumaGastos gastosVisibles = {[]}/>);
    // lo que esperamos del resultado que coincida con el snapshot
    expect(wrapper).toMatchSnapshot();
});

test('Deberia renderear bien el componente con un gasto', () =>{
    // Lo que se va renderear al usuario
    const wrapper = shallow(<DesplegadoSumaGastos gastosVisibles = {gastosPrueba2}/>);
    // lo que esperamos del resultado que coincida con el snapshot
    expect(wrapper).toMatchSnapshot();
});

