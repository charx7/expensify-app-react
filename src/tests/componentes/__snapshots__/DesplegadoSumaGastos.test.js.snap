// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deberia renderear bien el componente con un gasto 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DesplegadoSumaGastos
    gastosVisibles={
      Array [
        Object {
          "creadoEn": 0,
          "descripcion": "Gum",
          "id": "1",
          "monto": 195,
          "nota": "",
        },
        Object {
          "creadoEn": -345600000,
          "descripcion": "Renta",
          "id": "2",
          "monto": 1000,
          "nota": "",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Viendo 
        2
         gasto
        s 
        sumando: 
        $11.95
      </p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Viendo ",
          2,
          " gasto",
          "s ",
          "sumando: ",
          "$11.95",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Viendo ",
        2,
        " gasto",
        "s ",
        "sumando: ",
        "$11.95",
      ],
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Viendo 
          2
           gasto
          s 
          sumando: 
          $11.95
        </p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Viendo ",
            2,
            " gasto",
            "s ",
            "sumando: ",
            "$11.95",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Viendo ",
          2,
          " gasto",
          "s ",
          "sumando: ",
          "$11.95",
        ],
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Deberia renderear bien el componente con un gasto vacio 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <DesplegadoSumaGastos
    gastosVisibles={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        Viendo 
        0
         gasto
        s 
        sumando: 
        $0
      </p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          "Viendo ",
          0,
          " gasto",
          "s ",
          "sumando: ",
          "$0",
        ],
      },
      "ref": null,
      "rendered": Array [
        "Viendo ",
        0,
        " gasto",
        "s ",
        "sumando: ",
        "$0",
      ],
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          Viendo 
          0
           gasto
          s 
          sumando: 
          $0
        </p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Viendo ",
            0,
            " gasto",
            "s ",
            "sumando: ",
            "$0",
          ],
        },
        "ref": null,
        "rendered": Array [
          "Viendo ",
          0,
          " gasto",
          "s ",
          "sumando: ",
          "$0",
        ],
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
